File: Java/ApplicationLoggingUnitTests/src/net/alastairwyse/applicationloggingunittests/FileApplicationLoggerTests.java
/*
 * Copyright 2014 Alastair Wyse (http://www.oraclepermissiongenerator.net/methodinvocationremoting/)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.alastairwyse.applicationloggingunittests;

import org.junit.Before;
import org.junit.Test;
import static org.mockito.Mockito.*;
import net.alastairwyse.applicationlogging.*;
import net.alastairwyse.operatingsystemabstraction.*;

/**
 * Unit tests for class applicationlogging.FileApplicationLogger.
 * @author Alastair Wyse
 */
public class FileApplicationLoggerTests {
    private IFileWriter mockFileWriter;
    private FileApplicationLogger testFileApplicationLogger;
    
    @Before
    public void setUp() {
        mockFileWriter = mock(IFileWriter.class);
        testFileApplicationLogger = new FileApplicationLogger(LogLevel.Debug, ':', "  ", mockFileWriter);
    }
    
    @Test
    public void LogSuccessTests() throws Exception {
        String expectedStackTraceText = System.lineSeparator() + "  java.lang.Exception: Error connecting to /127.0.0.1:50000." + System.lineSeparator() + "  \tat net.alastairwyse.methodinvocationremoting.TcpRemoteSender.AttemptConnect(TcpRemoteSender.java:187)" + System.lineSeparator() + "  \tat net.alastairwyse.methodinvocationremoting.TcpRemoteSender.Connect(TcpRemoteSender.java:136)" + System.lineSeparator() + "  \tat net.alastairwyse.applicationloggingunittests.FileApplicationLoggerTests.LogSuccessTests(Test.java:135)" + System.lineSeparator();
        
        testFileApplicationLogger.Log(LogLevel.Critical, "Log Text 1.");
        testFileApplicationLogger.Log(123, LogLevel.Error, "Log Text 2.");
        testFileApplicationLogger.Log(LogLevel.Warning, "Log Text 3.", new CustomStackTraceException("Error connecting to /127.0.0.1:50000."));
        testFileApplicationLogger.Log(124, LogLevel.Information, "Log Text 4.", new CustomStackTraceException("Error connecting to /127.0.0.1:50000."));
        testFileApplicationLogger.Log(this, LogLevel.Critical, "Log Text 5.");
        testFileApplicationLogger.Log(this, 123, LogLevel.Error, "Log Text 6.");
        testFileApplicationLogger.Log(this, LogLevel.Warning, "Log Text 7.", new CustomStackTraceException("Error connecting to /127.0.0.1:50000."));
        testFileApplicationLogger.Log(this, 124, LogLevel.Information, "Log Text 8.", new CustomStackTraceException("Error connecting to /127.0.0.1:50000."));
        
        verify(mockFileWriter).write(contains(" CRITICAL : Log Text 1."));
        verify(mockFileWriter).write(contains(" Log Event Id = 123 : ERROR : Log Text 2."));
        verify(mockFileWriter).write(contains(" WARNING : Log Text 3." + expectedStackTraceText));
        verify(mockFileWriter).write(contains(" Log Event Id = 124 : Log Text 4." + expectedStackTraceText));
        verify(mockFileWriter).write(contains(" Source = FileApplicationLoggerTests : CRITICAL : Log Text 5."));
        verify(mockFileWriter).write(contains(" Source = FileApplicationLoggerTests : Log Event Id = 123 : ERROR : Log Text 6."));
        verify(mockFileWriter).write(contains(" Source = FileApplicationLoggerTests : WARNING : Log Text 7." + expectedStackTraceText));
        verify(mockFileWriter).write(contains(" Source = FileApplicationLoggerTests : Log Event Id = 124 : Log Text 8." + expectedStackTraceText));
        
        verify(mockFileWriter, times(8)).write(System.lineSeparator());
        verify(mockFileWriter, times(8)).flush();
        verifyNoMoreInteractions(mockFileWriter);
    }
    
    @Test
    public void LogBelowMinimumLogLevelNotLogged() throws Exception {
        testFileApplicationLogger = new FileApplicationLogger(LogLevel.Warning, ':', "  ", mockFileWriter);
        
        testFileApplicationLogger.Log(LogLevel.Information, "Log Text 1.");
        testFileApplicationLogger.Log(123, LogLevel.Information, "Log Text 2.");
        testFileApplicationLogger.Log(LogLevel.Information, "Log Text 3.", new Exception("Mocked Exception."));
        testFileApplicationLogger.Log(124, LogLevel.Information, "Log Text 4.", new Exception("Mocked Exception 2."));
        testFileApplicationLogger.Log(LogLevel.Debug, "Log Text 1.");
        testFileApplicationLogger.Log(123, LogLevel.Debug, "Log Text 2.");
        testFileApplicationLogger.Log(LogLevel.Debug, "Log Text 3.", new Exception("Mocked Exception."));
        testFileApplicationLogger.Log(124, LogLevel.Debug, "Log Text 4.", new Exception("Mocked Exception 2."));
        testFileApplicationLogger.Log(this, LogLevel.Information, "Log Text 5.");
        testFileApplicationLogger.Log(this, 123, LogLevel.Information, "Log Text 6.");
        testFileApplicationLogger.Log(this, LogLevel.Information, "Log Text 7.", new Exception("Mocked Exception 3."));
        testFileApplicationLogger.Log(this, 124, LogLevel.Information, "Log Text 8.", new Exception("Mocked Exception 4."));
        testFileApplicationLogger.Log(this, LogLevel.Debug, "Log Text 5.");
        testFileApplicationLogger.Log(this, 123, LogLevel.Debug, "Log Text 6.");
        testFileApplicationLogger.Log(this, LogLevel.Debug, "Log Text 7.", new Exception("Mocked Exception 3."));
        testFileApplicationLogger.Log(this, 124, LogLevel.Debug, "Log Text 8.", new Exception("Mocked Exception 4."));
        
        verifyZeroInteractions(mockFileWriter);
    }
    
    @Test
    public void CloseSuccessTest() throws Exception {
        testFileApplicationLogger.Close();
        
        verify(mockFileWriter).close();
        verifyNoMoreInteractions(mockFileWriter);
    }
}

File: C#/ApplicationLoggingUnitTests/FileApplicationLoggerTests.cs
/*
 * Copyright 2014 Alastair Wyse (http://www.oraclepermissiongenerator.net/methodinvocationremoting/)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Generic;
using System.Text;
using NUnit.Framework;
using NMock2;
using NMock2.Matchers;
using ApplicationLogging;
using OperatingSystemAbstraction;

namespace ApplicationLoggingUnitTests
{
    //******************************************************************************
    //
    // Class: FileApplicationLoggerTests
    //
    //******************************************************************************
    /// <summary>
    /// Unit tests for class ApplicationLogging.FileApplicationLogger.
    /// </summary>
    [TestFixture]
    public class FileApplicationLoggerTests
    {
        private Mockery mocks;
        private IStreamWriter mockStreamWriter;
        private FileApplicationLogger testFileApplicationLogger;

        [SetUp]
        protected void SetUp()
        {
            mocks = new Mockery();
            mockStreamWriter = mocks.NewMock<IStreamWriter>();
            testFileApplicationLogger = new FileApplicationLogger(LogLevel.Debug, ':', "  ", mockStreamWriter);
        }

        [Test]
        public void LogSuccessTests()
        {
            string expectedStackTraceText = Environment.NewLine + @"     at ApplicationLoggingUnitTests.LogSuccessTests() in C:\MethodInvocationRemoting\C#\ApplicationLoggingUnitTests\FileApplicationLoggerTests.cs:line 123" + Environment.NewLine + @"     at ApplicationLoggingUnitTests.LogSuccessTests() in C:\MethodInvocationRemoting\C#\ApplicationLoggingUnitTests\FileApplicationLoggerTests.cs:line 456";

            using (mocks.Ordered)
            {
                Expect.Once.On(mockStreamWriter).Method("WriteLine").With(new StringContainsMatcher(" CRITICAL : Log Text 1."));
                Expect.Once.On(mockStreamWriter).Method("Flush").WithNoArguments();
                Expect.Once.On(mockStreamWriter).Method("WriteLine").With(new StringContainsMatcher(" Log Event Id = 123 : ERROR : Log Text 2."));
                Expect.Once.On(mockStreamWriter).Method("Flush").WithNoArguments();
                Expect.Once.On(mockStreamWriter).Method("WriteLine").With(new StringContainsMatcher(" WARNING : Log Text 3." + Environment.NewLine + "  ApplicationLoggingUnitTests.CustomStackTraceException: Mocked Exception." + expectedStackTraceText));
                Expect.Once.On(mockStreamWriter).Method("Flush").WithNoArguments();
                Expect.Once.On(mockStreamWriter).Method("WriteLine").With(new StringContainsMatcher(" Log Event Id = 124 : Log Text 4." + Environment.NewLine + "  ApplicationLoggingUnitTests.CustomStackTraceException: Mocked Exception 2." + expectedStackTraceText));
                Expect.Once.On(mockStreamWriter).Method("Flush").WithNoArguments();
                Expect.Once.On(mockStreamWriter).Method("WriteLine").With(new StringContainsMatcher(" Source = FileApplicationLoggerTests : CRITICAL : Log Text 5."));
                Expect.Once.On(mockStreamWriter).Method("Flush").WithNoArguments();
                Expect.Once.On(mockStreamWriter).Method("WriteLine").With(new StringContainsMatcher(" Source = FileApplicationLoggerTests : Log Event Id = 123 : ERROR : Log Text 6."));
                Expect.Once.On(mockStreamWriter).Method("Flush").WithNoArguments();
                Expect.Once.On(mockStreamWriter).Method("WriteLine").With(new StringContainsMatcher(" Source = FileApplicationLoggerTests : WARNING : Log Text 7." + Environment.NewLine + "  ApplicationLoggingUnitTests.CustomStackTraceException: Mocked Exception 3." + expectedStackTraceText));
                Expect.Once.On(mockStreamWriter).Method("Flush").WithNoArguments();
                Expect.Once.On(mockStreamWriter).Method("WriteLine").With(new StringContainsMatcher(" Source = FileApplicationLoggerTests : Log Event Id = 124 : Log Text 8." + Environment.NewLine + "  ApplicationLoggingUnitTests.CustomStackTraceException: Mocked Exception 4." + expectedStackTraceText));
                Expect.Once.On(mockStreamWriter).Method("Flush").WithNoArguments();
            }
            testFileApplicationLogger.Log(LogLevel.Critical, "Log Text 1.");
            testFileApplicationLogger.Log(123, LogLevel.Error, "Log Text 2.");
            testFileApplicationLogger.Log(LogLevel.Warning, "Log Text 3.", new CustomStackTraceException("Mocked Exception."));
            testFileApplicationLogger.Log(124, LogLevel.Information, "Log Text 4.", new CustomStackTraceException("Mocked Exception 2."));
            testFileApplicationLogger.Log(this, LogLevel.Critical, "Log Text 5.");
            testFileApplicationLogger.Log(this, 123, LogLevel.Error, "Log Text 6.");
            testFileApplicationLogger.Log(this, LogLevel.Warning, "Log Text 7.", new CustomStackTraceException("Mocked Exception 3."));
            testFileApplicationLogger.Log(this, 124, LogLevel.Information, "Log Text 8.", new CustomStackTraceException("Mocked Exception 4."));

            mocks.VerifyAllExpectationsHaveBeenMet();
        }

        [Test]
        public void LogBelowMinimumLogLevelNotLogged()
        {
            testFileApplicationLogger = new FileApplicationLogger(LogLevel.Warning, ':', "  ", mockStreamWriter);

            Expect.Never.On(mockStreamWriter);

            testFileApplicationLogger.Log(LogLevel.Information, "Log Text 1.");
            testFileApplicationLogger.Log(123, LogLevel.Information, "Log Text 2.");
            testFileApplicationLogger.Log(LogLevel.Information, "Log Text 3.", new CustomStackTraceException("Mocked Exception."));
            testFileApplicationLogger.Log(124, LogLevel.Information, "Log Text 4.", new CustomStackTraceException("Mocked Exception 2."));
            testFileApplicationLogger.Log(LogLevel.Debug, "Log Text 1.");
            testFileApplicationLogger.Log(123, LogLevel.Debug, "Log Text 2.");
            testFileApplicationLogger.Log(LogLevel.Debug, "Log Text 3.", new CustomStackTraceException("Mocked Exception."));
            testFileApplicationLogger.Log(124, LogLevel.Debug, "Log Text 4.", new CustomStackTraceException("Mocked Exception 2."));
            testFileApplicationLogger.Log(this, LogLevel.Information, "Log Text 5.");
            testFileApplicationLogger.Log(this, 123, LogLevel.Information, "Log Text 6.");
            testFileApplicationLogger.Log(this, LogLevel.Information, "Log Text 7.", new CustomStackTraceException("Mocked Exception 3."));
            testFileApplicationLogger.Log(this, 124, LogLevel.Information, "Log Text 8.", new CustomStackTraceException("Mocked Exception 4."));
            testFileApplicationLogger.Log(this, LogLevel.Debug, "Log Text 5.");
            testFileApplicationLogger.Log(this, 123, LogLevel.Debug, "Log Text 6.");
            testFileApplicationLogger.Log(this, LogLevel.Debug, "Log Text 7.", new CustomStackTraceException("Mocked Exception 3."));
            testFileApplicationLogger.Log(this, 124, LogLevel.Debug, "Log Text 8.", new CustomStackTraceException("Mocked Exception 4."));
            mocks.VerifyAllExpectationsHaveBeenMet();
        }

        [Test]
        public void CloseSuccessTest()
        {
            using (mocks.Ordered)
            {
                Expect.Once.On(mockStreamWriter).Method("Close").WithNoArguments();
            }

            testFileApplicationLogger.Close();
            mocks.VerifyAllExpectationsHaveBeenMet();
        }
    }
}
